File: questApplication/src/main/java/ru/nekit/android/qls/quest/mediator/adapter/SquareImageAdapter.java
package ru.nekit.android.qls.quest.mediator.adapter;

import android.support.annotation.DrawableRes;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import java.util.List;

import ru.nekit.android.qls.R;
import ru.nekit.android.qls.quest.mediator.IContentContainerViewHolder;

public class SquareImageAdapter extends SquareItemAdapter<SquareImageAdapter.ImageViewHolder> {

    @LayoutRes
    private final int mImageLayoutResId;
    @NonNull
    private final List<?> mDataList;
    @DrawableRes
    private final List<Integer> mImageResourceIds;
    @NonNull
    private final View.OnClickListener mClickListener;

    public SquareImageAdapter(@LayoutRes int imageLayoutResId,
                              @NonNull List<?> dataList,
                              @NonNull List<Integer> imageResourceIds,
                              @NonNull View.OnClickListener clickListener) {
        mImageLayoutResId = imageLayoutResId;
        mImageResourceIds = imageResourceIds;
        mDataList = dataList;
        mClickListener = clickListener;
    }

    @Override
    public void onBindViewHolder(final ImageViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);
        View view = holder.itemView;
        if (mClickListener != null) {
            view.setOnClickListener(mClickListener);
        }
        view.setTag(mDataList == null ? null : mDataList.get(position));
        holder.getImageView().setImageResource(mImageResourceIds.get(position));
    }

    @Override
    public void onViewDetachedFromWindow(ImageViewHolder holder) {
        holder.itemView.setOnClickListener(null);
        super.onViewDetachedFromWindow(holder);
    }


    @Override
    public ImageViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return new ImageViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(mImageLayoutResId, parent, false));
    }

    @Override
    public int getItemCount() {
        return mImageResourceIds.size();
    }

    public static class ImageViewHolder extends RecyclerView.ViewHolder
            implements IContentContainerViewHolder {

        private final View mView, mImageView, mContentContainer;

        ImageViewHolder(@NonNull View view) {
            super(view);
            mView = view;
            mContentContainer = view.findViewById(R.id.container_content);
            mImageView = view.findViewById(R.id.view_image);
        }

        @NonNull
        public View getView() {
            return mView;
        }

        @NonNull
        @Override
        public View getContentContainer() {
            return mContentContainer;
        }

        @NonNull
        ImageView getImageView() {
            return (ImageView) mImageView;
        }
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/quest/view/adapter/SquareImageAdapter.kt
package ru.nekit.android.qls.quest.view.adapter

import android.support.annotation.DrawableRes
import android.support.annotation.LayoutRes
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.jakewharton.rxbinding2.view.clicks
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.Subject
import ru.nekit.android.qls.R
import ru.nekit.android.qls.quest.view.mediator.IContentContainerViewHolder

//ver 1.1
class SquareImageAdapter(@param:LayoutRes @field:LayoutRes
                         private val imageLayoutResId: Int,
                         private val dataList: List<*>,
                         @field:DrawableRes
                         private val imageResourceIdList: List<Int>,
                         answerPublisher: Subject<Any>) :
        SquareItemAdapter<SquareImageAdapter.ImageViewHolder>(answerPublisher) {

    override var disposableMap: MutableMap<String, CompositeDisposable> = HashMap()

    override fun onBindViewHolder(holder: ImageViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)
        holder.apply {
            itemView.apply {
                autoDispose {
                    clicks().map { itemView.tag }.subscribe({ answerPublisher.onNext(it) })
                }
                tag = dataList[position]
            }
            imageView.setImageResource(imageResourceIdList[position])
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ImageViewHolder =
            ImageViewHolder(LayoutInflater.from(parent.context)
                    .inflate(imageLayoutResId, parent, false))

    override fun getItemCount(): Int = imageResourceIdList.size

    class ImageViewHolder internal constructor(val view: View) : RecyclerView.ViewHolder(view),
            IContentContainerViewHolder {
        val imageView: ImageView = view.findViewById(R.id.view_image)
        override val contentContainer: View = view.findViewById(R.id.container_content)
    }
}